enhancements/monitoring/replace-prometheus-adapter-with-metrics-server.md

```diff
@@ -0,0 +1,261 @@
+
+  ```go
+  type ClusterMonitoringConfiguration struct {
+    ...
+    // `MetricsServerConfig` defines settings for the MetricsServer component.
+    MetricsServerConfig *MetricsServerConfig `json:"metricsServer,omitempty"`
+  }
+
+  // The `MetricsServerConfig` resource defines settings for the MetricsServer component.
+  type MetricsServerConfig struct {
+    // Defines the audit configuration used by the Metrics Server instance.
+    Audit *Audit `json:"audit,omitempty"`
+    // Defines the nodes on which the pods are scheduled.
+    NodeSelector map[string]string `json:"nodeSelector,omitempty"`
+    // Defines tolerations for the pods.
+    Tolerations []v1.Toleration `json:"tolerations,omitempty"`
+    // Defines resource requests and limits for the Metrics Server container.
+    Resources *v1.ResourceRequirements `json:"resources,omitempty"`
+    // Defines a pod's topology spread constraints.
+    TopologySpreadConstraints []v1.TopologySpreadConstraint `json:"topologySpreadConstraints,omitempty"`
+  }
+  ```
+
+The cluster monitoring operator will check the presence of the `MetricsServer` FeatureGate: if present, it will configure the Metrics
+Server resources in `openshift-monitoring` namespace and tears down the Prometheus Adapter resources once Metrics Server is fully functional. We will use the Prometheus Adapter settings to configure the Metrics server when the upgrade
+happens to avoid unpleasant surprises for users.
+
+### Test Plan
+
+The proposed solution will be tested with two suites:
+
+* Unit tests - running independently of each other and without side effects
+* End to end tests - in the same repository as the cluster-monitoring-operator. These tests will be executed against an actual Kubernetes cluster and are intended to verify end-to-end functionality of the entire system.
+
+E2E Test will test following:
+* Enables metrics-server
+* Verifies that the resource metrics API is working
+* During the test, it will ensure that there's no disruption of the resource metrics API.
+
+TechPreview jobs will be configured for the cluster monitoring operator and metrics-server repositories.
+
+### Graduation Criteria
+
+The main rationale behind adopting a phased approach is to do:
+- More testing
+- Conduct Peformance testing on larger cluster
+- Sufficient time for feedback
+
+* 4.15 TechPreview
+  * Notice in the release notes that prometheus-adapter is deprecated (should be repeated on every future release).
+* 4.16 GA, the `MetricsServer` feature gate will be enabled in the default feature set.
+
+#### Dev Preview -> Tech Preview
+
+N/A
+
+#### Tech Preview -> GA
+
+More testing (upgrade, downgrade, scale).
+* Sufficient time for feedback including signals from telemetry about the customer adoption
+  * We will be adding a Gauge Metric named `cluster_monitoring_operator_metrics_server_enabled` in [cmo](https://github.com/openshift/cluster-monitoring-operator/blob/master/pkg/metrics/metrics.go) to measure this.
```

Comment:
@simonpasquier I am wondering if we would need this now since we moved to featuregates and metrics server will enabled by default on tech preview? Shouldn't we able to use any existing metrics to see how many techpreview clusters created for 4.15?

Reply:
correct we don't need the extra metric in CMO.